crates/astria-sequencer/src/app.rs

```diff
@@ -426,24 +459,21 @@ impl App {
             );
         }
 
+        self.execution_results = Some(execution_results);
         (signed_txs, validated_txs)
     }
 
-    #[instrument(name = "App::finalize_block", skip_all)]
-    pub(crate) async fn finalize_block(
+    // sets up the state for execution of the block's transactions.
+    // set the current height and timestamp, and calls `begin_block` on all components.
+    //
+    // this *must* be called anytime before a block's txs are executed, whether it's
+    // during the proposal phase, or finalize_block phase.
+    async fn pre_execute_transactions(
         &mut self,
-        finalize_block: abci::request::FinalizeBlock,
-        storage: Storage,
-    ) -> anyhow::Result<abci::response::FinalizeBlock> {
-        use tendermint::{
-            abci::types::ExecTxResult,
-            block::Header,
-        };
-
-        use crate::transaction::InvalidNonce;
-
+        block_data: BlockData,
+    ) -> anyhow::Result<Vec<bytes::Bytes>> {
         let data_hash = astria_core::sequencerblock::v1alpha1::block::merkle_tree_from_data(
-            finalize_block.txs.iter().map(std::convert::AsRef::as_ref),
+            block_data.txs.iter().map(std::convert::AsRef::as_ref),
```

Comment:
If I'm reading things right, the generated `data_hash` could differ when called from `prepare_proposal()` than other places since it's using the unfiltered set of transactions vs filtered. I didn't see any of the begin_block() component implementations using it so I don't think it's going to cause problems as-is, but this could cause issues in the future.

Reply:
good point, this `data_hash` is only used in the `begin_block` struct below which is passed into the components, but none of them actually use the `data_hash` so this does nothing, i'll remove it!